File: day2/solution_p2.hs
parseMove str =
  let w = words str
      f = head w
      s = read . head $ tail w :: Int
   in case f of
        "forward" -> (\a b -> b * s + a, 0, s)
        "down" -> (const, s, 0)
        _ -> (const, - s, 0)

sumTuples (a, b, c) (fn, x, y) = (fn a b, b + x, c + y)

solution = do
  m <- map parseMove . lines <$> readFile "input.txt"
  let (u, _, v) = foldl sumTuples (0, 0, 0) m
  print $ show $ u * v

File: day2/solution_p2.rs
fn add(value: Option<i32>) -> Box<dyn Fn(i32, i32) -> i32> {
    value.map_or(Box::new(|a, _| (a)), |v| Box::new(move |a, b| (b * v + a)))
}

fn parse_move(line: &str) -> (Box<dyn Fn(i32, i32) -> i32>, i32, i32) {
    line.split_once(" ")
        .map(|(dir, mag)| {
            let value = mag.parse::<i32>().unwrap();
            match dir {
                "forward" => (add(Some(value)), 0, value),
                "down" => (add(None), value, 0),
                _ => (add(None), -value, 0i32),
            }
        })
        .unwrap()
}

pub fn solution() -> i32 {
    let (x, _, y) = std::fs::read_to_string("../input.txt")
        .unwrap()
        .lines()
        .map(|it| parse_move(it))
        .fold((0, 0, 0), |(a, b, c), (f, x, y)| (f(a, b), b + x, c + y));
    return x * y;
}
